### PR TITLE
enable generation of polynomial reconstruction coefficient up to rank 1 input

### PR DESCRIPTION
**Pull request by zishengye**
Previously, function applyFullPolynomialCoefficientsBasisToDataAllComponents only allows rank 0 input. I have extended it to allow rank 1 input. In this case, this function could be able to work for vector point samples as well.

### CODE REVIEW

**Comment by kuberry** [comment]
Hi @zishengye, are you still happy with the changes as reflected in this PR? Sorry that it has been a while between you requesting it and me finally having a chance to take a look at it.

**Comment by zishengye** [comment]
It's up to you @kuberry .

This part of work would help with 2d divergence free basis computation and adaptive refinement which needs function applyFullPolynomialCoefficientsBasisToDataAllComponents with vector field. This could make the project more complete. I am happy to see this part be added into the whole project.

However, one drawback of current code, which I do not have time to deal with, is that the function applyFullPolynomialCoefficientsBasisToDataAllComponents needs the number of batches to 1. If you have a solution to solve this, the deployment of adaptive refinement could be much more smooth in the future.

**Comment by kuberry** [comment]
Hi @zishengye, I'm almost through the code, just down to the part about applyFullPolynomialCoefficientsBasisToDataAllComponents. 

Would I be correct in understanding that you can now handle generating polynomial coefficients (for a scalar) by applying the solution in Coeffs to data that has more than one component (vector input)?

This covers the case of StaggeredEdgeIntegralSample which has a sampling type that takes vector input and a scalar coefficient basis (rank 0).

File: examples/GMLS_DivergenceFree.cpp [hunk 1]
```diff
@@ -476,13 +480,11 @@ bool all_passed = true;
         if(std::abs(actual_curl_vector[0] - GMLS_Curl_DivFree_VectorX) > failure_tolerance) {
             all_passed = false;
             std::cout << i << " Failed curl VectorX by: " << std::abs(actual_curl_vector[0] - GMLS_Curl_DivFree_VectorX) << std::endl;
-            if (dimension>1) {
+            if (dimension>2) {
```
[the remark this answers is not in the document]
Comment (zishengye) [reply examples/GMLS_DivergenceFree.cpp:483]: Well, for 2d problem, the curl operator only contains one component which is conflict with the current code. Also, it seems that the curl operator for 2d case is unfinished. You'd better check it.

**Comment by zishengye** [comment]
> Hi @zishengye, I'm almost through the code, just down to the part about applyFullPolynomialCoefficientsBasisToDataAllComponents.
> 
> Would I be correct in understanding that you can now handle generating polynomial coefficients (for a scalar) by applying the solution in Coeffs to data that has more than one component (vector input)?
> 
> This covers the case of StaggeredEdgeIntegralSample which has a sampling type that takes vector input and a scalar coefficient basis (rank 0).

I am not quite sure of it. I only tested with simple vector input like the case working with a divergence free basis.

**Comment by kuberry** [comment]
Hi @zishengye, thanks for the reply. Ok, sounds good. I was merging master into your branch to test some things, make minor changes, and push. 

Whenever this gets done, it always looks like you changed a bunch of files. If your remote branch is up-to-date (can you let me know?), I can do a force push with a rebased branch so that the differences are easier to see.

**Comment by zishengye** [comment]
My own branch has been far behind the current master head since you have added a lot of new features recently. It is almost the same as the one I made pull request here. I frozen my own branch for some other my ongoing projects. Sorry about that.

**Comment by kuberry** [comment]
@zishengye, thanks for your contributions to enabling polynomial coefficient calculation for bases where the _sampling_multiplier > 1. I just made a few more changes to enable _basis_multiplier > 1 as well (VectorTaylorPolynomial basis would qualify).

The evaluator is the most complicated part of this code, and I really appreciate you taking the time to understand how it works so that it can be improved. 

We can have a call sometime (Skype?) if you want to discuss batching and its effect on coefficient storage.

**Comment by zishengye** [comment]
@kuberry , I am happy to have a call over Skype. How about sometime next week? I think I could share some experience on using the evaluator.

**Pull request merged**